packages/core/src/modules/question-answer/services/QuestionAnswerService.ts

```diff
@@ -0,0 +1,250 @@
+      questionText: questionMessage.questionText,
+      questionDetail: questionMessage.questionDetail,
+      threadId: questionMessage.id,
+      connectionId: connectionId,
+      role: QuestionAnswerRole.Questioner,
+      signatureRequired: false,
+      state: QuestionAnswerState.QuestionSent,
+      validResponses: questionMessage.validResponses,
+    })
+
+    await this.questionAnswerRepository.save(questionAnswerRecord)
+
+    this.eventEmitter.emit<QuestionAnswerStateChangedEvent>({
+      type: QuestionAnswerEventTypes.QuestionAnswerStateChanged,
+      payload: { previousState: null, questionAnswerRecord },
+    })
+
+    return { questionMessage, questionAnswerRecord }
+  }
+
+  /**
+   * receive question message and create record for responder role
+   *
+   * @param messageContext the message context containing a question message
+   * @returns QuestionAnswer record
+   */
+  public async receiveQuestion(messageContext: InboundMessageContext<QuestionMessage>): Promise<QuestionAnswerRecord> {
+    const { message: questionMessage, connection } = messageContext
+
+    this.logger.debug(`Receiving question message with id ${questionMessage.id}`)
+
+    if (!connection) {
+      throw new AriesFrameworkError(`Connection for verkey ${messageContext.recipientVerkey} not found!`)
+    }
+
+    if (!connection.theirKey) {
+      throw new AriesFrameworkError(`Connection with verkey ${connection.verkey} has no recipient keys.`)
+    }
+
+    const questionAnswerRecord = await this.createRecord({
+      questionText: questionMessage.questionText,
+      questionDetail: questionMessage.questionDetail,
+      connectionId: connection?.id,
+      threadId: questionMessage.id,
+      role: QuestionAnswerRole.Responder,
+      signatureRequired: false,
+      state: QuestionAnswerState.QuestionReceived,
+      validResponses: questionMessage.validResponses,
+    })
+
+    await this.questionAnswerRepository.save(questionAnswerRecord)
+
+    this.eventEmitter.emit<QuestionAnswerStateChangedEvent>({
+      type: QuestionAnswerEventTypes.QuestionAnswerStateChanged,
+      payload: { previousState: null, questionAnswerRecord },
+    })
+
+    return questionAnswerRecord
+  }
+
+  /**
+   * create answer message, check that response is valid
+   *
+   * @param questionAnswerRecord record containing question and valid responses
+   * @param response response used in answer message
+   * @returns answer message and QuestionAnswer record
+   */
+  public async createAnswer(questionAnswerRecord: QuestionAnswerRecord, response: string) {
+    const answerMessage = new AnswerMessage({ response: response, threadId: questionAnswerRecord.threadId })
+
+    questionAnswerRecord.response = response
+
+    if (questionAnswerRecord.validResponses.some((e) => e.text === response)) {
+      await this.updateState(questionAnswerRecord, QuestionAnswerState.AnswerSent)
+    } else {
+      this.logger.error(`Response does not match valid responses`)
```

Comment:
Maybe we can throw an error here? It is not clear when calling this that the record won't get updated.